x11rb-protocol/benches/proto_connection.rs

```diff
@@ -0,0 +1,198 @@
+//! Benchmark the `x11rb_protocol::Connection` type's method, at varying levels of
+//! capacity.
+
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use std::mem::replace;
+use x11rb_protocol::connection::{Connection, ReplyFdKind};
+
+fn pow(i: i32, p: i32) -> i32 {
+    let mut result = 1;
+    for _ in 0..p {
+        result *= i;
+    }
+    result
+}
+
+fn enqueue_packet_test(c: &mut Criterion) {
+    // take the cartesian product of the following conditions:
+    // - the packet is an event, a reply, or an error
+    // - pending_events and pending_replies are empty, have one element, or have
+    //   many elements
+
+    enum PacketType {
+        Event,
+        Reply,
+        Error,
+    }
+
+    enum PacketCount {
+        Empty,
+        One,
+        Many,
+    }
+
+    use PacketCount::*;
+    use PacketType::*;
+
+    let mut group = c.benchmark_group("enqueue_packet");
+    for packet_ty in &[Event, Reply, Error] {
+        for packet_count in &[Empty, One, Many] {
+            let packet_ty_desc = match packet_ty {
+                Event => "event",
+                Reply => "reply",
+                Error => "error",
+            };
+
+            let packet_count_desc = match packet_count {
+                Empty => "no",
+                One => "one",
+                Many => "many",
+            };
+
+            let name = format!(
+                "enqueue_packet {} with {} packets",
+                packet_ty_desc, packet_count_desc
+            );
+            group.bench_function(name, |b| {
+                // generate a valid packet with the given first byte and sequence number
+                let mut seqno = 0u16;
+                let mut packet = move |ind: u8| {
+                    let our_seqno = seqno + 1;
+                    seqno += 1;
+
+                    let mut v = vec![0; 32];
+                    v[0] = ind;
+
+                    // copy our_seqno to bytes 3 and 4
+                    v[2] = (our_seqno >> 8) as u8;
+                    v[3] = our_seqno as u8;
```

Comment:
This assumes little endian and would break on a big endian system, wouldn't it?

I think this needs `v[2..4].copy_from_slice(our_seqno.to_ne_bytes());`